### ISSUE TITLE
Yolov5 and inference with batch_size > 1

### ISSUE THREAD

**Issue by s-nesterov** [wang-xinyu/tensorrtx#487]
Author, thank you for your work! A fairly detailed tutorial on converting weights. I managed to convert the weights to tensorrt format, trained on my dataset. The increase in the speed of the model was almost 2 times, which is just great! However, it is saddening to use the model in the tesorrt format only with batch_size=1 (we are talking about the python wrapper). Tell me, how can I use a model in tensorrt format with batch>1?

Thank you in advance

**Comment by freedenS** [comment]
You can try to set BATCH_SIZE  > 1 ( yolov5.cpp line 13)

**Comment by s-nesterov** [comment]
> You can try to set BATCH_SIZE > 1 ( yolov5.cpp line 13)

Yes, I tried to do that. However, with inference, I get a strange result - in the batch there are predictions only for the first image, everything else is zero

**Comment by freedenS** [comment]
Is BATCH_SIZE > 1 both with serialization and deserialization?

**Comment by wang-xinyu** [comment]
The .cpp can work well with multi-batch. But .py cannot currently.

**Comment by vsantosu** [comment]
Having the same issue here... Batch works for CPP but not python. Maybe the input shape? The input shape of the resulting model is:

(3, 608, 608), instead of (8, 3, 608, 608) for a batch size of 8...

**Comment by wang-xinyu** [comment]
@vsantosu  the yolov5_trt.py does not support multi-batch right now. You can use cpp or revise .py yourself.

**Comment by vsantosu** [comment]
Hello folks,

First, thanks all who contributed to this repo, pretty amazing work! I will try to break up all my findings regarding this issue. The truth is, the problem is way more complex than just adapting they Python TensorRT inference. Here are my findings:

TL;DR: It is not possible to do batch inference with the current serialized engine build in the native CPP process(Unless I'm missing something). Here is why:

1. I changed the CPP BATCH_SIZE = 8, and exported the engine. The benchmarks seems to indicate the the engine is in fact doing batch inference...

2. I adapted the python code to do batch inference. NOTE: I use this code for other TensorRT models and the batch inference works perfectly fine:

```
    def infer(self, input_original, write_results=False):
        # Restore
        stream = self.stream
        context = self.context
        engine = self.engine
        host_inputs = self.host_inputs
        cuda_inputs = self.cuda_inputs
        host_outputs = self.host_outputs
        cuda_outputs = self.cuda_outputs
        bindings = self.bindings
        batch_size = self.batch_size
        
        # the batch array
        input_batch = np.zeros((batch_size, 3, 608, 608))
        input_hw = []

        # compute total time
        then_total = time.time()
        # preprocess all images
        then = time.time()
        for i, image_raw in enumerate(input_original):
            # Do image preprocess
            input_image, image_raw, origin_h, origin_w = self.preprocess_image(
                image_raw
            )
            # set batch of images here
            input_batch[i] = input_image
            input_hw.append((origin_h, origin_w))

        # compute preprocessing for the batch
        now = time.time()
        print(f"precessesing: {now-then} secs")

        # Make self the active context, pushing it on top of the context stack.
        self.cfx.push()
        then = time.time()
        # Copy input image to host buffer
        np.copyto(host_inputs[0], input_batch.ravel())
        # Transfer input data  to the GPU.
        cuda.memcpy_htod_async(cuda_inputs[0], host_inputs[0], stream)
        # Run inference.
        context.execute_async(bindings=bindings, stream_handle=stream.handle)
        # Transfer predictions back from the GPU.
        cuda.memcpy_dtoh_async(host_outputs[0], cuda_outputs[0], stream)
        # Synchronize the stream
        stream.synchronize()
        now = time.time()
        print(f"inference: {now-then} secs")

        # Remove any context from the top of the context stack, deactivating it.
        self.cfx.pop()
        # Here we use the first row of output in that batch_size = 1

        output = host_outputs[0]
        bpr = int(output.shape[0]/batch_size)
        outputs = output.reshape((batch_size, bpr))

        # the result array
        results = []
        # Do postprocess
        then = time.time()
        for i in range(len(input_original)):
            # extract values
            inf_res = outputs[i]
            origin_h, origin_w = input_hw[i]
            # post process image
            result_boxes, result_scores, result_classid = self.post_process(
                inf_res, origin_h, origin_w
            )
            print(result_boxes)
            # push result
            results.append((result_boxes, result_scores, result_classid))
        
        now = time.time()
        print(f"post processing: {now-then} secs")

        now_total = time.time()
        print(f"TOTAL: {now_total-then_total} secs")

        # if write results, the spill into disk
        if write_results:
            self.result_to_file(results, input_original)

    def result_to_file(self, results, input_original):
        # Draw rectangles and labels on the original image
        for i in range(len(results)):
            # extract values
            result_boxes, result_scores, result_classid = results[i]
            image_raw =  input_original[i]

            for j in range(len(result_boxes)):
                cls_idx = int(result_classid[j])
                # if not a person continue
                if cls_idx != 0:
                    continue
                box = result_boxes[j]
                plot_one_box(
                    box,
                    image_raw,
                    label="{}:{:.2f}".format(
                        categories[cls_idx], result_scores[j]
                    ),
                )

            # _, filename = os.path.split(input_image_path)
            filename = f"{i}.jpg"
            save_name = os.path.join("build", "output_" + filename)
            # 　Save image
            cv2.imwrite(save_name, image_raw)
            # cv2.imshow('test', image_raw)
            # cv2.waitKey(0)

```

3. Here is where things gets weird... The inference seems to be doing batch, but **only the first image spills out results**. If you do batch=1, the output is 1*3*608*608, you the reshape and it is all good. If you do batch=8, the the result output is: 8*3*608*608
you then reshape and process the output one by one, but again, only the first one provide results, the rest of the output tensor is 0.

4. I tested many things like checking that the cuda copy from host to GPU was correct, and indeed was copying correctly. Then I dig more into the problem and I found the same issue in other places: [NVIDIA Forum](https://forums.developer.nvidia.com/t/batch-inference-wrong-in-python-api/112183/12)

5. If we inspect the input and output shapes of the serialized engine we get the following:

```

        # display input and output shape
        input_shape = engine.get_binding_shape(0)
        output_shape = engine.get_binding_shape(1)
        batch_size = engine.max_batch_size
        # print engine info
        print(f"engine input: {input_shape}, output: {output_shape}, batch size: {batch_size}")

```
output is:

```
engine input: (3, 608, 608), output: (6001, 1, 1), batch size: 8
```
6. What I understand from this is that the batch size must be explicit and since the serialized engine input is: (3, 608, 608), it will only pick the first input tensor from the batch, and will spill out the full batch but only the first results will be present.

7. So, the network needs to be serialized with explicit batch. I jumped into the C++ code and try to do this by changing the network definition to explicit shape:

```
    const auto explicitBatch = 1U << static_cast<uint32_t>(NetworkDefinitionCreationFlag::kEXPLICIT_BATCH);
    INetworkDefinition* network = builder->createNetworkV2(explicitBatch);
```

The problem here is that now the tensor sizes does not match and I start getting errors.

From this point forward, I don't know what to do, the CPP network definition must be modified correctly to include explicit batch size and the serialized engine to be imported in python with input size like (8, 3, 608, 608) and output size like (48008, 1, 1).
Currently it is: (3, 608, 608), output: (6001, 1, 1), and it is why you only get one result even if doing batch.

PD: I have done this with other models by just exporting from pytorch to onnx with explicit batch size:

```
    torch.onnx.export(model, x, onnx_path,opset_version = 11, input_names=['input'],
                  output_names=['output'], dynamic_axes={'input':{0:'batch_size'}, 'output':{0:'batch_size'}}, export_params=True)

```

Then using **trtexec**, I can fix the batch sizes of the ONNX model like:


```
# set the batch size
BATCH_SIZE=6
MEMORY_SIZE=2024

trtexec --explicitBatch \
        --onnx=model.onnx \
        --saveEngine=model.trt \
        --workspace="$MEMORY_SIZE" \
        --shapes=input:"$BATCH_SIZE"x3x256x256 \
        --fp16 
```

Sorry for the blog post. I hope this helps and somebody finds the solution, this is by far the best implementation of Yolov5 for TensorRT.

**Comment by wang-xinyu** [comment]
@vsantosu Thanks for your post.

I'm going to explain a bit about it.

The last layer of this yolov5 tensorrt model is a yololayer plugin, the src code is https://github.com/wang-xinyu/tensorrtx/blob/master/yolov5/yololayer.cu.

The `yololayer` output dimension is 

yololayer.cu:
```
        int totalsize = mMaxOutObject * sizeof(Detection) / sizeof(float);
        return Dims3(totalsize + 1, 1, 1);
```

by default, mMaxOutObject = 1000, so dims will be (6001, 1, 1).

And it is using implicit batchsize, and the batchsize is limited by this line when build engine

yolov5.cpp:
```
builder->setMaxBatchSize(maxBatchSize);
```

So if you set BATCH_SIZE = 1, you can only feed (1, 3, H, W) to model. output will be (1, 6001, 1, 1)

If you set BATCH_SIZE = 8, you can feed (b, 3, H, W) , b is >=1 and <=8. And the output will be (b, 6001, 1, 1).

When doing inference for multi-batch, you need pay attention to doInfence(), that is you must feed batch data, and must also fetch batch data from it.

yolov5.cpp
```
void doInference(IExecutionContext& context, cudaStream_t& stream, void **buffers, float* input, float* output, int batchSize) {
    // DMA input batch data to device, infer on the batch asynchronously, and DMA output back to host
    CUDA_CHECK(cudaMemcpyAsync(buffers[0], input, batchSize * 3 * INPUT_H * INPUT_W * sizeof(float), cudaMemcpyHostToDevice, stream));
    context.enqueue(batchSize, buffers, stream, nullptr);
    CUDA_CHECK(cudaMemcpyAsync(output, buffers[1], batchSize * OUTPUT_SIZE * sizeof(float), cudaMemcpyDeviceToHost, stream));
    cudaStreamSynchronize(stream);
}
```

And finally, for how to decode the output (b, 6001, 1, 1), you need also check the yolov5.cpp.

**Comment by s-nesterov** [comment]
If someone succeeds, I will be grateful for something like a guide with explanations

**Comment by wang-xinyu** [comment]
@vsantosu @s-nesterov you can check the latest codes for yolov5 batch inference. closing this one.

**Comment by vsantosu** [comment]
@wang-xinyu , I just tried the new version, I can confirm it works like a charm!! Thanks so much for the effort you put into this, I really appreciate it!

**Comment by s-nesterov** [comment]
Thank you so much for your work! Everything really works!
